data_steward/analytics/cdr_ops/combined.py

```diff
@@ -0,0 +1,479 @@
+# there is no way to specify `combined`.
+
+query = f'''
+WITH schema_opts AS
+(SELECT 
+ catalog_name
+,schema_name
+,SPLIT( REPLACE( REPLACE( REPLACE( 
+        SUBSTR(option_value, 2, LENGTH(option_value)-4), 
+        "STRUCT(\\"", '' ), 
+        "\\", \\"", ":" ), 
+        "\\"), ", "," )
+      ) AS labels
+FROM `INFORMATION_SCHEMA.SCHEMATA_OPTIONS` 
+WHERE OPTION_NAME="labels")
+SELECT 
+ catalog_name
+,schema_name
+,STRUCT(
+    ARRAY_AGG(STRUCT(SPLIT(un_labels, ":")[OFFSET(0)]         AS name,
+    SPLIT(un_labels, ":")[OFFSET(1)] AS value) RESPECT NULLS) AS label_object
+ ) AS label
+FROM schema_opts
+LEFT JOIN UNNEST(labels) AS un_labels
+WHERE schema_name='{DATASET_ID}'
+GROUP BY 1, 2
+'''
+q(query)
+
+# ## Invalid concept prevalence
+# EHR submissions may provide source `concept_id`s which we assume would be
+# successfully mapped to standard `concept_id`s. This compares the frequency
+# of invalid values in required concept_id fields per hpo between the current
+# combined dataset and a baseline combined dataset. The frequency of invalid
+# values in this dataset should generally be non-increasing compared to baseline.
+# If the prevalence of invalid concepts is increasing, EHR Ops should be alerted.
+
+FQN_INVALID_CONCEPT = 'curation_sandbox.invalid_concept'
+
+query = f'''
+CREATE OR REPLACE TABLE `{FQN_INVALID_CONCEPT}` 
+(dataset_id  STRING NOT NULL OPTIONS(description="Identifies the dataset")
+,table_name  STRING NOT NULL OPTIONS(description="Identifies the table")
+,column_name STRING NOT NULL OPTIONS(description="Identifies the concept_id field")
+,src_hpo_id  STRING NOT NULL OPTIONS(description="Identifies the HPO that supplied the records")
+,concept_id  INT64           OPTIONS(description="A value that is not in the associated concept table (can be NULL)")
+,row_count   INT64  NOT NULL)
+OPTIONS
+(description='Count of invalid values provided in required concept_id fields for multiple datasets.'
+,expiration_timestamp=TIMESTAMP_ADD(CURRENT_TIMESTAMP(), INTERVAL 7 DAY) )
+'''
+q(query)
+
+tpl = JINJA_ENV.from_string('''
+-- Construct a query that finds all rows in all tables --
+-- where concept_id fields have an invalid value and reports --
+-- the row count per site --
+DECLARE query DEFAULT (
+    SELECT 
+ STRING_AGG('SELECT ' 
+     || '"{{DATASET_ID}}"'  || ' AS dataset_id '
+     || ',"' || table_name  || '"  AS table_name '
+     || ',"' || column_name || '" AS column_name '
+     || ', m.src_hpo_id AS src_hpo_id '
+     || ',' || column_name  || ' AS concept_id '
+     || ', COUNT(1) AS row_count '
+     -- domain table joined with its mapping table --
+     || 'FROM `' || table_schema || '.' || table_name || '` t'
+     || ' JOIN `' || table_schema || '.' || '_mapping_' || table_name || '` m'
+     || '  USING (' || c.table_name || '_id) '
+     || ' LEFT JOIN `{{DATASET_ID}}.concept` c '
+     || '  ON ' || column_name || ' = c.concept_id '
+     
+     -- invalid concept_id --
```

Comment:
Thank you for these comments!  Makes reading the SQL much more manageable.

Reply:
I spend more time than one might think on trying to make SQL intelligible. I'm so glad you noticed!